CAP-0065 - Reusable module cache

graydon opened the pull request:
This is the stellar-core side of a [soroban change to surface the module cache for reuse](https://github.com/stellar/rs-soroban-env/pull/1506).

On the core side we:
  1. Add a new mechanism to scan the live BL snapshot for contracts alone (which is a small part of the BL)
  2. Add a new wrapper type `SorobanModuleCache` to the core-side Rust code, which holds a `soroban_env_host::ModuleCache` for each host protocol version we support caching modules for (there is currently only one but there will be more in the future)
  3. Also add a `CoreCompilationContext` type to `contract.rs` which carries a `Budget` and logs errors to the core console logging system. This is sufficient to allow operating the `soroban_env_host::ModuleCache` from outside the `Host`.
  4. Pass a `SorobanModuleCache` into the host function invocation path that core calls during transactions.
  5. Store a `SorobanModuleCache` in the `LedgerManagerImpl` that is long-lived, spans ledgers.
  6. Add a utility class `SharedModuleCacheCompiler` that does a multithreaded load-all-contracts / populate-the-module-cache, and call this on startup when the `LedgerManagerImpl` restores its LCL.
  7. Add a couple command-line helpers that list and forcibly compile all contracts (we might choose to remove these, they were helpful during debugging but at this point perhaps redundant)

The main things left to do here are:

 - [x] Maybe move all this to the soon-to-be-added `p23` soroban submodule
 - [x] Decide how we're going to handle growing the cache with new uploads and wire that up
 - [x] Similarly decide how we're going to handle expiry and restoration events [as discussed in CAP0062](https://github.com/stellar/stellar-protocol/blob/master/core/cap-0062.md) and wire that up too
 - [x] Write some tests if I can think of any more specific than just "everything existing still works"
 - [x] Write a CAP describing this

I think that's .. kinda it? The reusable module cache is just "not passed in" on p22 and "passed in" on p23, so it should just start working at the p23 boundary.

dmkozh reviewed (commented):
IIUC this still needs to be rebased on top of p23 module and have the host change merged, so I haven't looked too closely at the rust parts (IIUC currently we just don't pass the module cache to the host at all).

SirTyson reviewed (changes requested):
I think this generally looks pretty good, the producer/compiler thread divide is a good idea! I think the producer thread needs some changes though.

To maintain the cache, I think we need to
1. Compile all live contract modules on startup.
2. Compile new contract modules during upload/restore.
3. Evict entries from the cache when the underlying WASM is evicted via State Archival eviction scans.

I think decoupling cache gc from eviction events is going to be expensive. If you have some background task that checks if a given module is live or not, it will have to read from many levels of the BL to determine if whatever BucketEntry it's looking at is the most up to date. Evicting from the cache when we do state archival evictions will remove this additional read amplification (since the archival scan has to this this multi level search already) and is simpler to maintain cache validity too.

The drawback to this is intial cache generation is a little more expensive, as we're limited to a single producer thread that has to iterate throught the BL in-order and keep track of seen keys. If we don't have a proper gc, we can't add any modules that have already been evited since they would cause a memory leak.

Looking at startup as a whole, we have a bunch of tasks that are BL disk read dominated, namely Bucket Apply, BucketIndex, and p23's upcoming Soroban state cache. Bucket Index can process all Buckets in parallel, but Bucket Apply, Soroban State cache, and Module Cache all require a single thread iterating the BL in order due to the outdated keys issue (in the future we could do this in parallel where each level marks it's "last seen key" and lower levels can't make progress beyond all their parents last seen key, but that's too invloved for v1).

Given that we're adding a bunch of work on the startup path and Horizon/RPC has indicated a need for faster startup times in the past, I think it makes sense to condense the Bucket Apply, Soroban state cache population, and Module cache producer thread into a single Work that makes a one shot pass on the BucketList. Especially in a captive-core instance, which we still run in our infra on EBS last I checked, I assume we're going to be disk bound even with the compilation step, so if we do compilation in the same pass as BucketApply we might just get it for free.

I don't think this needs to be in 23.0 (other than the memory leak issue), but if we have this in mind and make the initial version a little more friendly with the other Work tasks that happen on startup, it'll be easier to optimize this later.

dmkozh commented on src/ledger/SharedModuleCacheCompiler.cpp:
```diff
@@ -0,0 +1,161 @@
+
+    for (auto thread = 0; thread < self->mNumThreads; ++thread)
+    {
+        mApp.postOnBackgroundThread(
```
Aren't we using the same low priority threads here, by using postOnBackgroundThread? That's my concern. But my main concern is the flag semantics - since background threads are indeed meant to be IO bound, the value of the setting doesn't correlate with what we actually want here (roughly matching the physical cores available).

graydon replied:
Eh, I don't think this actually matters much but it's an easy enough change, and exploring the "right" number of threads actually took me in an interesting direction! More than 6 starts to slow down, presumably because of the gratuitous amount of locking and notifying in my lousy implementation coupled with the very short runtime of each of the compilations (like on the order of 100us). Probably this could be tuned to be more efficient, but at 6 threads it compiles all 700 contracts in 170ms, which seems fine! I added a config parameter to control this also.

graydon commented:
Updated to include:
  - Rebased
  - Adapted to simplified module cache in recently-merged upstream
  - Moved soroban changes to p23 submodule
  - Incremental expiry and addition of new contracts after ledger close
  - Less-chatty logging
  - Added support for populating module cache for side-by-side testing with `SOROBAN_TEST_EXTRA_PROTOCOL=23`

graydon commented on src/ledger/SharedModuleCacheCompiler.h:
```diff
@@ -0,0 +1,59 @@
+    bool mLoadedAll{false};
+    size_t mBytesLoaded{0};
+    size_t mBytesCompiled{0};
+    std::vector<uint32_t> mLedgerVersions;
```
It varies depending on the protocol that's actually running. If we're on protocol 25 there's no point compiling protocols 23 and 24. Protocols only upgrade.

dmkozh replied:
I see, thanks. I'm still find it a bit puzzling that the interface suggests that it's possible to have gaps in protocol versions (vs a range), but probably that doesn't matter too much.

graydon commented on src/ledger/SharedModuleCacheCompiler.h:
```diff
@@ -0,0 +1,59 @@
+    stellar::SearchableSnapshotConstPtr mSnap;
+    std::deque<xdr::xvector<uint8_t>> mWasms;
+
+    const size_t mNumThreads;
```
fixed.

graydon commented on src/ledger/SharedModuleCacheCompiler.h:
```diff
@@ -0,0 +1,59 @@
+    std::deque<xdr::xvector<uint8_t>> mWasms;
+
+    const size_t mNumThreads;
+    const size_t MAX_MEM_BYTES = 10 * 1024 * 1024;
```
fixed (also renamed for clarity)

graydon commented on src/ledger/SharedModuleCacheCompiler.h:
```diff
@@ -0,0 +1,59 @@
+
+// This class encapsulates a multithreaded strategy for loading contracts
+// out of the database (on one thread) and compiling them (on N-1 others).
+class SharedModuleCacheCompiler
```
Oh weird. I have no idea! That was a mistake. Fixed.

graydon commented on src/ledger/SharedModuleCacheCompiler.h:
```diff
@@ -0,0 +1,59 @@
+    std::deque<xdr::xvector<uint8_t>> mWasms;
+
+    const size_t mNumThreads;
+    const size_t MAX_MEM_BYTES = 10 * 1024 * 1024;
```
It's for buffering-while-loading (like it's the allowed gap that the load thread is allowed to get ahead of the consuming/compiling threads). But I can bump it up to 100, sure, don't care.

graydon commented on src/main/CommandLine.cpp:
```diff
@@ -1945,6 +1964,11 @@ handleCommandLine(int argc, char* const* argv)
          {"test", "execute test suite", runTest},
          {"apply-load", "run apply time load test", runApplyLoad},
 #endif
+         {"list-contracts",
```
Oh, actually these are artifacts of development, we don't really need to keep either of these commands. They're trivial enough that if I ever needed them again I'd just write them again. Will remove entirely for now.

graydon commented on src/rust/src/contract.rs:
```diff
@@ -630,3 +636,92 @@ pub(crate) fn can_parse_transaction(xdr: &CxxBuf, depth_limit: u32) -> bool {
+        let unlimited_budget = Budget::try_from_configs(
+            u64::MAX,
+            u64::MAX,
+            ContractCostParams(vec![].try_into().unwrap()),
```
I thought so too! But no, the code that initializes the cost model is driven by the input. It iterates over the input updating the dimension's default cost model, as it goes. No input, no updates. The nice thing about this is it's also relatively future proof: if you initialize a newer soroban's dimension from an older (say on-chain) cost parameters vector, it'll just overwrite the prefix it specified.

We could of course change this, or make it more explicit. But I feel like that's kinda a separate change, involves deciding on a different API. For the time being I felt ok using this.

graydon commented on src/ledger/SharedModuleCacheCompiler.h:
```diff
@@ -0,0 +1,59 @@
+#pragma once
+// Copyright 2024 Stellar Development Foundation and contributors. Licensed
```
Fixed

graydon commented on src/ledger/SharedModuleCacheCompiler.cpp:
```diff
@@ -0,0 +1,161 @@
+#include "ledger/SharedModuleCacheCompiler.h"
```
Fixed

graydon commented on src/ledger/SharedModuleCacheCompiler.h:
```diff
@@ -0,0 +1,59 @@
+// This class encapsulates a multithreaded strategy for loading contracts
+// out of the database (on one thread) and compiling them (on N-1 others).
+class SharedModuleCacheCompiler
+    : public std::enable_shared_from_this<SharedModuleCacheCompiler>
```
I don't think so:

- The point is to actually run in parallel, for speedup. Work does not run in parallel, it's for long-running async tasks that are interleaved on the main thread.
- In general Work is very code-heavy and awkward. If something is isolated enough / within our risk tolerance enough to run on a thread, IMO it's much less work to just put it on one.
- This task only takes a half second on my machine for all contracts on chain currently; it's 4s of CPU time but again: parallelism! It may grow to as much as a few seconds, but..
- If we have a long-running thread we _do_ need to be able to "exit more gracefully", IMO we should just make the shutdown flag atomic and threadsafe and have the thread poll it periodically. Much less clunky than a full Work class.

I mean, I think Work classes still have a place (especially when we're talking C++ and it's comparatively dangerous / impossible to safely add threads to some main-thread-spanning activities) but I think in this case, and several like it the affected memory graph is all threadsafe (the cache and the bucket snapshots) so it's probably ok to thread.

graydon commented on src/bucket/BucketSnapshot.h:
```diff
@@ -87,6 +87,10 @@ class LiveBucketSnapshot : public BucketSnapshotBase<LiveBucket>
+
+    // Scans contract code entries in the bucket.
+    Loop
+    scanForContractCode(std::function<Loop(LedgerEntry const&)> callback) const;
```
In case the lambda wants to stop early (eg. it's looking for a specific contract, or an error occurred, or whatever)

graydon replied:
Oh and re: thread contention, I don't think it's a huge deal again because it's just not that long. We could cut the thread count down if you want a margin of safety, but when I test it it just scrolls past instantaneously like you don't even see it happening.

graydon commented on src/ledger/SharedModuleCacheCompiler.cpp:
```diff
@@ -0,0 +1,161 @@
+    LOG_INFO(DEFAULT_LOG,
+             "Launching 1 loading and {} compiling background threads",
+             mNumThreads);
+    mApp.postOnBackgroundThread(
```
Oh, huh. So .. I guess I should:
  - Pass the full bucketentry to the callback
  - Keep track, in the compilation loop, of whether I've seen the hash before (live or dead) 
  - Only compile live entries _I haven't seen before_ (live or dead)

I _think_ that should do it?

graydon replied:
Ok, did this. Seems to work!

graydon commented:
Updated with fix for test failure when running with background close, as well as all review comments addressed (I think!)

dmkozh reviewed (commented):
LGTM, but CI is still failing

SirTyson reviewed (changes requested):
I think there's a bug in the `ApplyBuckets` work path. Currently, we use the BucketList snapshot to compile WASM. Problem is, we call `compileAllContractsInLedger` in `AssumeStateWork`. After calling `app.getBucketManager().assumeState`, the raw BucketList is up to date, but we don't actually update the BucketList snapshots until after `AssumeStateWork` is down (see `LedgerManagerImpl.cpp:392`). I'm pretty sure we'd currently only compile state based on the genesis ledger when we hit the apply bucket path.

graydon commented:
> I think there's a bug in the `ApplyBuckets` work path. Currently, we use the BucketList snapshot to compile WASM. Problem is, we call `compileAllContractsInLedger` in `AssumeStateWork`. After calling `app.getBucketManager().assumeState`, the raw BucketList is up to date, but we don't actually update the BucketList snapshots until after `AssumeStateWork` is down (see `LedgerManagerImpl.cpp:392`). I'm pretty sure we'd currently only compile state based on the genesis ledger when we hit the apply bucket path.

Ah! You're right.

That path is actually only there to try to make compilation happen "post catchup", even though as it stands it is also triggered on the startup `AssumeState` as well. Incorrectly (as you mention) _and_ redundantly with the _explicit_ call to `compileAllContractsInLedger` below, at the bottom of `LedgerManagerImpl::loadLastKnownLedger`.

I think the fix here is to remove the call to `compileAllContractsInLedger` from `AssumeStateWork` altogether, and just do it in `LedgerApplyManager` when it decides the `CatchupWork` is complete. Then there won't be a redundant (and wrong!) compile happening during startup, and there _will_ be a non-redundant (and hopefully right!) compile happening during catchup.

graydon commented:
> Hmm I think we still have to compile before catch-up though (or at least before applying ledgers in catch-up)? I assume (perhaps incorrectly) that we have asserts that compilation is always cached and will remove the functionality to lazily compile modules during tx invocation in p23. TX replay would break these invariants.

Hmmmm maybe! I am unclear on where we should compile during catchup, then. Perhaps in `LedgerManagerImpl::setLastClosedLedger`, at the end?

graydon commented:
(In terms of asserts: there is no assert in the code currently that "every contract is cached" before we call a txn; fortunately or unfortunately, soroban will not-especially-gracefully degrade to making throwaway modules as needed. We could put an assert on the core side, and I guess we should since it represents a bug-that-turns-into-a-performance-hazard)

SirTyson commented:
> Hmmmm maybe! I am unclear on where we should compile during catchup, then. Perhaps in LedgerManagerImpl::setLastClosedLedger, at the end?

Yeah, that makes sense to me. Compilation is basically part of assuming the current ledger state, so seems like a reasonable place to put it.

graydon commented:
> Yeah, that makes sense to me. Compilation is basically part of assuming the current ledger state, so seems like a reasonable place to put it.

Done.

dmkozh commented on src/ledger/LedgerManagerImpl.cpp:
```diff
@@ -1822,6 +1971,8 @@ LedgerManagerImpl::transferLedgerEntriesToBucketList(
     ltx.getAllEntries(initEntries, liveEntries, deadEntries);
     if (blEnabled)
     {
+        addAnyContractsToModuleCache(lh.ledgerVersion, initEntries);
```
I'm not sure what is supposed to happen in the edge case scenario where a contract gets uploaded and called in two different transactions within the same ledger. Are we going to charge the full parse cost for the second call? Then it's likely going to fail due to exceeding the resources (because simulation assumes the contracts are in the cache). That's probably ok, since this is not something that's going to happen often. However, I just wanted to make sure that nothing blows up in the host.

graydon replied:
You're asking a few questions at once here, but I'll try to answer them all :)

1. The upload op _always_ charges a full parse, it doesn't look in (or immediately modify) the module cache at all. So all the following cases charge for a full parse, both in simulation and in execution:
  - 1 txset with 1 tx uploading a new contract
  - 1 txset with 1 tx uploading an existing-on-chain contract
  - 1 txset with 2 txs uploading the same contract
  - 2 txsets, each with 1 tx uploading the same contract (this is the same as the existing-on-chain case)
2. Executing a contract _usually_ won't charge a full parse for it, _unless_ it was missing from the txset's prestate:
  - 1 txset with 1 tx that uploads a new contract, 1 that runs it: second tx is charged a full parse
  - 1 txset with 1 tx that uploads an existing contract, 1 that runs it: second tx is not charged (contract existed in prestate)
  - Any other pattern of upload and-and-execute of contracts that exist in prestate: execution is not charged

The addition of entries _to_ the module cache happens only at txset boundaries, consolidating "all the new contracts that showed up in this txset". If they were already present, extra adds are idempotent.

Does that answer?

dmkozh replied:
Yeah, this all makes perfect sense, thanks.

dmkozh commented on src/ledger/test/LedgerTests.cpp:
```diff
@@ -45,3 +51,70 @@ TEST_CASE("cannot close ledger with unsupported ledger version", "[ledger]")
+    return true;
+}
+
+TEST_CASE("reusable module cache", "[ledger]")
```
This test is fine, but I think we need more integration-style tests (probably belonging to `InvokeHostFunctionOpFrameTests`). Here are some scenarios I have in mind:
- Start app in p22, upload and call a contract, record the instructions used. Then upgrade to p23, call the contract again and ensure the instructions went down
- Upload and call the contract in the same ledger in p23 (we can enforce the correct order of txs for this), ensure the instructions consumed are as expected (IIUC charged for full parse)
- Evict and restore a contract in p23 (within the same ledger, and within different ledgers), ensure instructions are consumed as expected (IIUC charged for full parse in the first case, not charged in the second)
I'm not sure if we have to block the PR on this if we're reasonably confident in p22 behavior, but I think we definitely need these for p23.

dmkozh replied:
I think the coverage looks good now, thanks!

dmkozh commented on src/ledger/LedgerManagerImpl.cpp:
```diff
@@ -1822,6 +1971,8 @@ LedgerManagerImpl::transferLedgerEntriesToBucketList(
     if (blEnabled)
     {
+        addAnyContractsToModuleCache(lh.ledgerVersion, initEntries);
+        addAnyContractsToModuleCache(lh.ledgerVersion, liveEntries);
```
Is this expected to be a no-op now, or do we have the restored entries in `liveEntries`?

graydon replied:
I don't know -- I figured it'd be fairly future-proof to just assume both live and init _might_ carry contracts. @SirTyson do you know off hand?

dmkozh replied:
Yeah, I think it shouldn't matter much, I just found this a bit suspicious, given that contract code must be immutable. But this shouldn't hurt either.

graydon commented on src/main/AppConnector.cpp:
```diff
@@ -108,10 +109,19 @@ AppConnector::getConfig() const
+AppConnector::getModuleCache()
 {
-    return *mModuleCache;
+    std::lock_guard<std::mutex> guard(mModuleCacheMutex);
```
Oh, huh. I think I forgot `<shared_mutex>` even exists. Yeah, sure.

graydon commented on src/catchup/AssumeStateWork.cpp:
```diff
@@ -9,6 +9,7 @@
 #include "crypto/Hex.h"
 #include "history/HistoryArchive.h"
 #include "invariant/InvariantManager.h"
+#include "ledger/LedgerManager.h"
```
Fixed.

graydon commented on src/transactions/test/InvokeHostFunctionTests.cpp:
```diff
@@ -4983,13 +5005,9 @@ TEST_CASE("Module cache cost with restore gaps", "[tx][soroban]")
-        auto tx2 = makeAddTx(contract, 700'000, A);
-        auto tx3 = makeAddTx(contract, 800'000, B);
+        // won't be populated and we need more instructions.
+        auto tx2 = makeAddTx(contract, 400'000, A);
```
whoops, fixed.

graydon commented on src/transactions/test/InvokeHostFunctionTests.cpp:
```diff
@@ -4740,3 +4741,278 @@ TEST_CASE("contract constructor support", "[tx][soroban]")
+TEST_CASE("Module cache across protocol versions", "[tx][soroban][modulecache]")
+{
+
+    if (protocolVersionIsBefore(Config::CURRENT_LEDGER_PROTOCOL_VERSION,
```
Oh, sure. Fixed.

graydon commented:
Rebased around @SirTyson's latest BL changes, repaired tests, integrated bits of residual feedback, squashed and cleaned up dead/redundant states, and picked out what I _think_ were all the build-breaking bugs.

The last one (in 1bb1b8dc2dcc7346b41db63ede1183421f4337e0) is one I suspect @SirTyson might disagree with; I do not like it either, and am very open to other options here! In general I do not understand, despite a lot of code-reading, why there are _static_ members of `LoadGenerator` rather than just member variables.

The presence of the statics means that global state of the `LoadGenerator` _carries forward_ from one test in the unit testsuite to another. This in turn means that depending on how tests are partitioned by the test runner (which changes as we add tests to the testsuite) tests that rely on that state break or un-break (in particular the `REQUIRE_THROWS_WITH` check in testcase `"generate soroban load"` of `LoadGeneratorTests.cpp` breaks or un-breaks depending on test order).

I think this is not ideal! Resetting the `LoadGenerator` statics on construction, as I'm doing in this change, seems like the least we could do. And seems (when I test it locally) to be enough to get tests passing. For now. But it also seems wrong. Like: why have static state at all? Is there no way to get rid of it?

dmkozh commented:
> In general I do not understand, despite a lot of code-reading, why there are static members of LoadGenerator rather than just member variables.

Yeah, I couldn't immediately tell either. Moreover, we seem to reset that 'soroban state' quite in a few places already. I suspect this works around some very particular scenarios that could just be fixed by preserving the loadgen instance in-between loadgen calls.

graydon commented:
> > In general I do not understand, despite a lot of code-reading, why there are static members of LoadGenerator rather than just member variables.
> 
> Yeah, I couldn't immediately tell either. Moreover, we seem to reset that 'soroban state' quite in a few places already. I suspect this works around some very particular scenarios that could just be fixed by preserving the loadgen instance in-between loadgen calls.

I talked to @SirTyson about this this morning and he thinks it was just a misunderstanding -- like when it was written he was somehow thinking that `LoadGenerator` wasn't long-lived or something. He believes it should be fine to remove.

After that discussion, I removed the `static` qualifiers and at least locally on my system the tests all pass. Pushed update with that change. I guess if it works, we can just see if we rediscover some reason for those qualifiers in the future?

SirTyson commented on src/ledger/LedgerManagerImpl.cpp:
```diff
@@ -1872,6 +2027,75 @@ LedgerManagerImpl::ledgerClosed(
+LedgerManagerImpl::evictFromModuleCache(uint32_t ledgerVersion,
+                                        EvictedStateVectors const& evictedState)
+{
+    std::vector<Hash> keys;
```
Nit: TracyZone

graydon replied:
done

SirTyson commented on src/ledger/LedgerManagerImpl.cpp:
```diff
@@ -1872,6 +2027,75 @@ LedgerManagerImpl::ledgerClosed(
+LedgerManagerImpl::addAnyContractsToModuleCache(
+    uint32_t ledgerVersion, std::vector<LedgerEntry> const& le)
+{
+    for (auto const& e : le)
```
Nit: TracyZone

graydon replied:
done

SirTyson commented on src/ledger/LedgerManagerImpl.cpp:
```diff
@@ -560,6 +593,118 @@ LedgerManagerImpl::getSorobanMetrics()
 
+::rust::Box<rust_bridge::SorobanModuleCache>
+LedgerManagerImpl::getModuleCache()
+{
```
Nit: I'd be interested in adding tracy zones to these functions. Especially as the ledger grows, might be interesting to see how we block on these calls.

graydon replied:
yep yep yep

SirTyson commented on src/main/AppConnector.h:
```diff
@@ -26,6 +28,12 @@ class AppConnector
+    // cache handle point to the same, shared, threadsafe module cache. It may
+    // periodically be replaced by a delete/rebuild cycle in the LedgerManager.
+    // This is always done under a mutex.
+    rust::Box<rust_bridge::SorobanModuleCache> mModuleCache;
```
I think this goes against the original design intention of the app connector. `AppConnector` is originally intended to ba a lightweight "threadsafe" wrapper around App. Since `App` has many main-thread only functions, it's dangerous to pass around, so we have a handful of functions in `AppConnector` that we've carefully audited and know are thread safe. Having state changes that contract, and makes me think the mModuleCache and `mModuleCacheMutex` should be moved to `LedgerManager`, and `AppConnector` just has a pass through to `LedgerManager` for those calls.

graydon replied:
I did try this initially but I then convinced myself that it wasn't correct, due to potential skews between the kicking-off of cache recompilation (which starts on main thread from the LCL-advancing step of `ledgerClosed`) and the point where cache recompilation _ends_ (on the close thread, just before execution) and the point where the cache is _used_ (on the close thread, repeatedly on each transaction).

Like I think the close thread can queue up multiple LCL-advancing callbacks for the main thread, and so any one of those on the main thread could decide to recompile the cache, making it change mid way through the close thread using it. Which wouldn't necessarily be a data race -- we mutex it -- but it would risk a nondeterministic miss or something, if the cache switches between two txs.

But now, re-reading it, to be honest I can't convince myself of _what_ it should do, because I can't tell which functions run on which threads without working backwards from first principles each time. This is one of the things that really worried (and continue to worry) me about non-main-thread closing in general: I can't clearly reason about which variables are advancing when, and there's no way to enforce separation between them.

For the time being we should pause landing this until I can re-derive a _crystal clear_ explanation, state-partition and lifecycle, and diagram it, and possibly figure out some way of actually enforcing it. I am really frustrated and really sorry.

SirTyson replied:
Hmmm, why do we need to kick of the recompilation in the main thread? I agree the story gets a little confusing if we start bouncing between main and background threads, but I don't think that's necessary here.

We know that the cache is stricly an apply-time cache (or should be an apply time cache), barring startup. There's no reason I can see that the cache then shouldn't just be completely managed by the apply thread. I think from a thread perspective, recompilation is similar to background eviction, where it's launched and joined on the apply thread. We can join the compilation right before we start apply (in the apply thread), then kick it off at the same place we call eviction scan (also in the apply thread). While we haven't techincally advanced LCL yet, the apply thread still has access to all upgraded network settings and protocol upgrades at the point at which we kick off the eviction scan, since we just finished applying those upgrades, so there shouldn't be any issues of stale snapshots or inconsistencies. We can guarantee that nothing has happened that could invalidate the cache in between it's start and finish, as only the apply thread can do that, and we only start/join compilation at the start and end of apply thread.

SirTyson replied:
With this approach we wouldn't even need locks. I believe Brett introduced the concept of a "thread assert" in #4637, where you can `assertIsMain()` and `assertIsApplyThread()`, which I imagine would be sufficient safety wise.

graydon replied:
@SirTyson Ok I've adapted this to the post-LedgerManager-refactoring PR (https://github.com/stellar/stellar-core/pull/4647) and rebased on master. Sure enough the additional structure was enough to make it fairly trivially clear which snapshots to use, when, and why.

The previous iteration of the PR was _mostly_ (or perhaps entirely) correct already, in the sense of compiling the right snapshots, but it was doing so more by accident than by intent. Or perhaps if I'm being very generous to myself: I had figured out the right place to do the recompiles already but forgotten why they were right, and was having trouble reconstructing the rationale. It's quite a bit more explicit now.

SirTyson commented on src/simulation/LoadGenerator.h:
```diff
@@ -282,11 +282,11 @@ class LoadGenerator
-    inline static UnorderedSet<LedgerKey> mContractInstanceKeys = {};
-    inline static std::optional<LedgerKey> mCodeKey = std::nullopt;
-    inline static uint64_t mContactOverheadBytes = 0;
+    UnorderedSet<LedgerKey> mContractInstanceKeys = {};
```
Thank you for fixing the sins of my past...

SirTyson commented:
LGTM! I'd like to see some tracy zones on some of the performance sensitive functions (although to be fair these may already exist rust side, I didn't look too closely on non-C++ stuff). Also a minor refactoring point on `AppConnector`

SirTyson reviewed (dismissed):
LGTM! I'd like to see some tracy zones on some of the performance sensitive functions (although to be fair these may already exist rust side, I didn't look too closely on non-C++ stuff). Also a minor refactoring point on AppConnector

graydon commented:
@SirTyson I believe this is ready to go again, though .. if we're holding back until post-release, it can continue to wait. But I'm not worried about its snapshot logic anymore!

dmkozh commented on src/transactions/test/InvokeHostFunctionTests.cpp:
```diff
@@ -4914,3 +4915,343 @@ TEST_CASE("contract constructor support", "[tx][soroban]")
+}
+
+static bool
+wasms_are_cached(Application& app, std::vector<Hash> const& wasms)
```
nit: `wasmsAreCached`

graydon replied:
fixed

graydon commented on src/main/AppConnector.h:
```diff
@@ -26,6 +28,12 @@ class AppConnector
+    // periodically be replaced by a delete/rebuild cycle in the LedgerManager.
+    // This is always done under a mutex.
+    rust::Box<rust_bridge::SorobanModuleCache> mModuleCache;
+    std::shared_mutex mModuleCacheMutex;
```
Oh, huh, actually so is the cached module cache itself! Residue of a previous iteration. Nice catch.

graydon commented on src/bucket/BucketSnapshot.cpp:
```diff
@@ -318,6 +318,43 @@ LiveBucketSnapshot::scanForEviction(
 }
 
+// Scans contract code entries in the bucket.
+Loop
```
This is the function to scan _one_ bucket, but it has to inform a higher level function whether to continue scanning to the _next_ bucket or not. It's passed to loopAllBuckets in SearchableLiveBucketListSnapshot::scanForContractCode -- the standard interface to loopAllBuckets involves passing this token outward from the callback through the per-bucket helper to loopAllBuckets. In practice in our current usage (in this PR) the innermost callback always passes INCOMPLETE, but different query patterns could change that and the one-bucket helper (LiveBucketSnapshot::scanForContractCode) should still plumb the value through.

graydon commented on src/ledger/SharedModuleCacheCompiler.cpp:
```diff
@@ -0,0 +1,225 @@
+
+    mThreads.emplace_back(std::thread([this]() {
+        ZoneScopedN("load wasm contracts");
+        std::set<Hash> seenContracts;
```
Oh, sure.

graydon commented on src/ledger/SharedModuleCacheCompiler.cpp:
```diff
@@ -0,0 +1,225 @@
+    mTotalCompileTime += dur_us;
+    mBytesCompiled += wasm.size();
+    wasm.clear();
+    mHaveSpace.notify_all();
```
It is intentional but the rationale isn't bulletproof. Basically I was struggling to get atomicity among all the checks, wakeup conditions and termination conditions, so I decided to make more of the functions assume the lock is held across them. I added a comment to clarify. I'm not sure if there's a better approach but I couldn't figure one out.

dmkozh commented on src/main/AppConnector.h:
```diff
@@ -4,6 +4,8 @@
 #include "main/Config.h"
 #include "medida/metrics_registry.h"
+#include "rust/RustBridge.h"
+#include <shared_mutex>
```
unused include

graydon replied:
Fixed
